src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs

```diff
@@ -0,0 +1,215 @@
+// <copyright file="PrometheusCollectionManager.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using OpenTelemetry.Metrics;
+
+namespace OpenTelemetry.Exporter.Prometheus
+{
+    internal sealed class PrometheusCollectionManager
+    {
+        private readonly PrometheusExporter exporter;
+        private readonly int scrapeResponseCacheDurationInMilliseconds;
+        private readonly Func<Batch<Metric>, ExportResult> onCollectRef;
+        private byte[] buffer = new byte[85000]; // encourage the object to live in LOH (large object heap)
+        private int globalLockState;
+        private ArraySegment<byte> previousDataView;
+        private DateTime? previousDataViewExpirationAtUtc;
+        private int readerCount;
+        private bool collectionRunning;
+        private TaskCompletionSource<ArraySegment<byte>> collectionTcs;
+
+        public PrometheusCollectionManager(PrometheusExporter exporter)
+        {
+            this.exporter = exporter;
+            this.scrapeResponseCacheDurationInMilliseconds = this.exporter.Options.ScrapeResponseCacheDurationInMilliseconds;
+            this.onCollectRef = this.OnCollect;
+        }
+
+#if NETCOREAPP3_1_OR_GREATER
+        public ValueTask<ArraySegment<byte>> EnterCollect()
+#else
+        public Task<ArraySegment<byte>> EnterCollect()
+#endif
+        {
+            this.EnterGlobalLock();
+
+            // If we are within 10 seconds of the last successful collect, return the previous view.
+            if (this.previousDataViewExpirationAtUtc.HasValue && this.previousDataViewExpirationAtUtc >= DateTime.UtcNow)
+            {
+                Interlocked.Increment(ref this.readerCount);
+                this.ExitGlobalLock();
+#if NETCOREAPP3_1_OR_GREATER
+                return new ValueTask<ArraySegment<byte>>(this.previousDataView);
```

Comment:
nit: for troubleshooting purposes, i think it'd be good to add some logging to indicate whether we re-used previous view, or triggered new collection.

Reply:
OK I was working on a log message but I'll hold off and do this as a follow-up. I like the idea of using the header.